File: rest-service/src/main/java/mb/lib/model/EmptyBlastConfig.java
package mb.lib.model;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.veupathdb.lib.blast.BlastConfig;
import org.veupathdb.lib.blast.BlastTool;
import org.veupathdb.lib.blast.consts.Flag;
import org.veupathdb.lib.blast.consts.Key;
import org.veupathdb.lib.blast.field.*;

import java.util.function.Function;

/**
 * Empty Blast Config
 * <p>
 * Represents an invalid blast configuration that may be left over in various
 * places if there is an issue with the config serialization.
 */
public class EmptyBlastConfig implements BlastConfig {
  protected final ObjectNode raw;

  protected JsonNode tmp;

  public EmptyBlastConfig(ObjectNode raw) {
    this.raw = raw;
  }

  @Override
  public BlastTool getTool() {
    return get(Key.Tool, BlastTool::fromJSON);
  }

  @Override
  public Boolean getShortHelp() {
    return getBoolean(Flag.ShortHelp);
  }

  @Override
  public void setShortHelp(Boolean aBoolean) {
    raw.put(Flag.ShortHelp, aBoolean);
  }

  @Override
  public Boolean getLongHelp() {
    return getBoolean(Flag.LongHelp);
  }

  @Override
  public void setLongHelp(Boolean aBoolean) {
    raw.put(Flag.LongHelp, aBoolean);
  }

  @Override
  public Boolean getVersion() {
    return getBoolean(Flag.Version);
  }

  @Override
  public void setVersion(Boolean aBoolean) {
    raw.put(Flag.Version, aBoolean);
  }

  @Override
  public OutFile getOutFile() {
    return fromString(Flag.OutFile, OutFile::new);
  }

  @Override
  public void setOutFile(String s) {
    raw.set(Flag.OutFile, new OutFile(s).toJSON());
  }

  @Override
  public OutFormat getOutFormat() {
    return get(Flag.OutFormat, OutFormat::fromJSON);
  }

  @Override
  public void setOutFormat(OutFormat outFormat) {
    raw.set(Flag.OutFormat, outFormat.toJSON());
  }

  @Override
  public Boolean getShowGIs() {
    return getBoolean(Flag.ShowGIs);
  }

  @Override
  public void setShowGIs(Boolean aBoolean) {
    raw.put(Flag.ShowGIs, aBoolean);
  }

  @Override
  public NumDescriptions getNumDescriptions() {
    return fromLong(Flag.NumDescriptions, NumDescriptions::new);
  }

  @Override
  public void setNumDescriptions(Long aLong) {
    raw.put(Flag.NumDescriptions, aLong);
  }

  @Override
  public NumAlignments getNumAlignments() {
    return fromLong(Flag.NumAlignments, NumAlignments::new);
  }

  @Override
  public void setNumAlignments(Long aLong) {
    raw.put(Flag.NumAlignments, aLong);
  }

  @Override
  public LineLength getLineLength() {
    return get(Flag.LineLength, JsonNode::intValue, LineLength::new);
  }

  @Override
  public void setLineLength(Integer integer) {
    raw.put(Flag.LineLength, integer);
  }

  @Override
  public Boolean getHTML() {
    return getBoolean(Flag.HTML);
  }

  @Override
  public void setHTML(Boolean aBoolean) {
    raw.put(Flag.HTML, aBoolean);
  }

  @Override
  public HitSorting getSortHits() {
    return get(Flag.SortHits, HitSorting::fromJSON);
  }

  @Override
  public void setSortHits(HitSorting hitSorting) {
    raw.set(Flag.SortHits, hitSorting.toJSON());
  }

  @Override
  public HSPSorting getSortHSPs() {
    return get(Flag.SortHSPs, HSPSorting::fromJSON);
  }

  @Override
  public void setSortHSPs(HSPSorting hspSorting) {
    raw.set(Flag.SortHSPs, hspSorting.toJSON());
  }

  @Override
  public MaxTargetSeqs getMaxTargetSequences() {
    return fromLong(Flag.MaxTargetSequences, MaxTargetSeqs::new);
  }

  @Override
  public void setMaxTargetSequences(Long aLong) {
    raw.put(Flag.MaxTargetSequences, aLong);
  }

  @Override
  public Boolean getParseDefLines() {
    return getBoolean(Flag.ParseDefLines);
  }

  @Override
  public void setParseDefLines(Boolean aBoolean) {
    raw.put(Flag.ParseDefLines, aBoolean);
  }

  //
  //
  // Internal
  //
  //

  protected <T> T fromLong(String key, Function<Long, T> fn) {
    return toTmp(key) && !tmp.isNull() ? fn.apply(tmp.longValue()) : null;
  }

  protected <T> T fromString(String key, Function<String, T> fn) {
    return toTmp(key) && !tmp.isNull() ? fn.apply(tmp.textValue()) : null;
  }

  protected Long getLong(String key) {
    return toTmp(key) && !tmp.isNull() ? tmp.longValue() : null;
  }

  protected Boolean getBoolean(String key) {
    return toTmp(key) && !tmp.isNull() ? tmp.booleanValue() : null;
  }

  protected String getString(String key) {
    return toTmp(key) && !tmp.isNull() ? tmp.textValue() : null;
  }

  protected <T> T get(String key, Function<JsonNode, T> fn) {
    return toTmp(key) && !tmp.isNull() ? fn.apply(tmp) : null;
  }

  protected <T, I> T get(String key, Function<JsonNode, I> g, Function<I, T> m) {
    return toTmp(key) && !tmp.isNull() ? m.apply(g.apply(tmp)) : null;
  }

  protected boolean toTmp(String key) {
    return (tmp = raw.get(key)) != null;
  }
}

File: rest-service/src/main/kotlin/mb/lib/model/EmptyBlastConfig.kt
package mb.lib.model

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.ObjectNode
import org.veupathdb.lib.blast.BlastConfig
import org.veupathdb.lib.blast.BlastTool
import org.veupathdb.lib.blast.consts.Flag
import org.veupathdb.lib.blast.consts.Key
import org.veupathdb.lib.blast.field.*

/**
 * Empty Blast Config
 * <p>
 * Represents an invalid blast configuration that may be left over in various
 * places if there is an issue with the config serialization.
 */
open class EmptyBlastConfig(protected val raw: ObjectNode): BlastConfig {

  protected var tmp: JsonNode? = null

  override val tool: BlastTool
    get() = get(Key.Tool, BlastTool::fromJSON)!!

  override var shortHelp: Boolean?
    get() = getBoolean(Flag.ShortHelp)
    set(v) {
      raw.put(Flag.ShortHelp, v)
    }

  override var longHelp: Boolean?
    get() = getBoolean(Flag.LongHelp)
    set(v) {
      raw.put(Flag.LongHelp, v)
    }

  override var version: Boolean?
    get() = getBoolean(Flag.Version)
    set(v) {
      raw.put(Flag.Version, v)
    }

  override var outFile: OutFile?
    get() = fromString(Flag.OutFile, ::OutFile)
    set(v) {
      raw.set<JsonNode>(Flag.OutFile, v?.toJSON())
    }

  override var outFormat: OutFormat?
    get() = get(Flag.OutFormat, OutFormat::fromJSON)
    set(v) {
      raw.set<JsonNode>(Flag.OutFormat, v?.toJSON())
    }

  override var showGIs: Boolean?
    get() = getBoolean(Flag.ShowGIs)
    set(v) {
      raw.put(Flag.ShowGIs, v)
    }

  override var numDescriptions
    get() = fromLong(Flag.NumDescriptions, ::NumDescriptions)
    set(v) {
      raw.put(Flag.NumDescriptions, v?.value)
    }

  override var numAlignments
    get() = fromLong(Flag.NumAlignments, ::NumAlignments)
    set(v) {
      raw.put(Flag.NumAlignments, v?.value)
    }

  override var lineLength: LineLength?
    get() = fromInt(Flag.LineLength, ::LineLength)
    set(v) {
      raw.put(Flag.LineLength, v?.value)
    }

  override var html: Boolean?
    get() = TODO("Not yet implemented")
    set(v) {
      raw.put(Flag.HTML, v)
    }

  override var sortHits: HitSorting?
    get() = get(Flag.SortHits, HitSorting::fromJSON)
    set(v) {
      raw.set<JsonNode>(Flag.SortHits, v?.toJSON())
    }

  override var sortHSPs: HSPSorting?
    get() = get(Flag.SortHSPs, HSPSorting::fromJSON)
    set(v) {
      raw.set<JsonNode>(Flag.SortHSPs, v?.toJSON())
    }

  override var maxTargetSequences: MaxTargetSeqs?
    get() = fromLong(Flag.MaxTargetSequences, ::MaxTargetSeqs)
    set(v) {
      raw.put(Flag.MaxTargetSequences, v?.value)
    }

  override var parseDefLines: Boolean?
    get() = getBoolean(Flag.ParseDefLines)
    set(v) {
      raw.put(Flag.ParseDefLines, v)
    }

  //
  //
  // Internal
  //
  //

  protected inline fun <T> fromInt(key: String, fn: (Int) -> T) =
    if (toTmp(key) && !tmp!!.isNull) fn(tmp!!.intValue()) else null

  protected inline fun <T> fromLong(key: String, fn: (Long) -> T) =
    if (toTmp(key) && !tmp!!.isNull) fn(tmp!!.longValue()) else null

  protected inline fun <T> fromString(key: String, fn: (String) -> T) =
    if (toTmp(key) && !tmp!!.isNull) fn(tmp!!.textValue()) else null

  protected inline fun getLong(key: String) =
    if (toTmp(key) && !tmp!!.isNull) tmp!!.longValue() else null

  protected inline fun getBoolean(key: String) =
    if (toTmp(key) && !tmp!!.isNull) tmp!!.booleanValue() else null

  protected inline fun getString(key: String) =
    if (toTmp(key) && !tmp!!.isNull) tmp!!.textValue() else null

  protected inline fun <T> get(key: String, fn: (JsonNode) -> T) =
    if (toTmp(key) && !tmp!!.isNull) fn(tmp!!) else null

  protected inline fun <T, I> get(key: String, g: (JsonNode) -> I, m: (I) -> T) =
    if (toTmp(key) && !tmp!!.isNull) m(g(tmp!!)) else null

  protected fun toTmp(key: String): Boolean {
    tmp = raw.get(key)
    return tmp != null
  }
}
